drivers/net/ethernet/intel/idpf/idpf_aux.c

```diff
@@ -0,0 +1,327 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/* Copyright (C) 2025 Intel Corporation */
+
+#include "idpf.h"
+#include "idpf_virtchnl.h"
+#include "idpf_aux_idc.h"
+
+static DEFINE_IDA(idpf_aux_dev_ida);
+
+/**
+ * idpf_aux_init_device_params - Initialize auxiliary device params
+ * @adapter: IDPF main private structure
+ * @eth_dev: Auxiliary (Ethernet) dev info
+ */
+static void
+idpf_aux_init_device_params(struct idpf_adapter *adapter,
+			    struct idpf_auxiliary_device *eth_dev)
+{
+	/* crc info */
+	eth_dev->aux_info.caps.crc_enable = true;
+
+	/* Queue Info */
+	eth_dev->aux_info.caps.q_info.max_rxq =
+		le16_to_cpu(adapter->caps.max_rx_q);
+	eth_dev->aux_info.caps.q_info.max_txq =
+		le16_to_cpu(adapter->caps.max_tx_q);
+	eth_dev->aux_info.caps.q_info.max_bufq =
+		le16_to_cpu(adapter->caps.max_rx_bufq);
+	eth_dev->aux_info.caps.q_info.max_complq =
+		le16_to_cpu(adapter->caps.max_tx_complq);
+
+	/* Caps */
+	eth_dev->aux_info.caps.csum_caps = adapter->caps.csum_caps;
+	eth_dev->aux_info.caps.seg_caps = adapter->caps.seg_caps;
+	eth_dev->aux_info.caps.hsplit_caps =
+		adapter->caps.hsplit_caps;
+	eth_dev->aux_info.caps.rsc_caps = adapter->caps.rsc_caps;
+	eth_dev->aux_info.caps.rss_caps = adapter->caps.rss_caps;
+	eth_dev->aux_info.caps.other_caps = adapter->caps.other_caps;
+	eth_dev->aux_info.caps.max_tx_hdr_size = adapter->caps.max_tx_hdr_size;
+	eth_dev->aux_info.caps.max_sg_bufs_per_tx_pkt =
+		adapter->caps.max_sg_bufs_per_tx_pkt;
+	eth_dev->aux_info.caps.min_sso_packet_len =
+		adapter->caps.min_sso_packet_len;
+}
+
+/**
+ * idpf_aux_release - Release IDPF aux device associated with auxiliary device
+ * @device: pointer to the device
+ */
+static void idpf_aux_release(struct device *device)
+{
+	struct auxiliary_device *adev = to_auxiliary_dev(device);
+	struct idpf_auxiliary_device *aux_dev;
+
+	aux_dev = container_of(adev, struct idpf_auxiliary_device,
+			       adev);
+	ida_free(&idpf_aux_dev_ida, aux_dev->adev.id);
+	kfree(aux_dev);
+}
+
+/**
+ * idpf_aux_alloc_dev - allocate one auxiliary device
+ * @adapter: IDPF main private structure
+ * @index: auxiliary device index witin the same function
+ * @vport_type: Vport type
+ *
+ * Returns allocated structure
+ */
+static struct idpf_auxiliary_device *
+idpf_aux_alloc_dev(struct idpf_adapter *adapter, unsigned int index,
```

Comment:
We will need an abstraction for idfp_adapter as ixd won't have that. What part of adapter is needed? For sure pdev pointer, but any other specific data?
 
 Maybe sth like aux_dev will fit here (idpf_aux_dev following comnvention). Corresponding driver will fill this structure based on their adapter types and pass it to alloc/free functions. What do you think?

Reply:
I thought about opaque API for aux creation to not double the same code in ixd,. As you wrote it can be handy to call some kind of _create_eth() (from idpf_eth.c). Will check if it is possible when implementing ixd side, thanks.